src/lib/dnssd/minimal_mdns/core/RecordWriter.cpp

```diff
@@ -0,0 +1,186 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "RecordWriter.h"
+
+namespace mdns {
+namespace Minimal {
+
+SerializedQNameIterator RecordWriter::PreviousName(size_t index) const
+{
+    if (index >= kMaxCachedReferences)
+    {
+        return SerializedQNameIterator();
+    }
+
+    uint16_t offset = mPreviousQNames[index];
+    if (offset == kInvalidOffset)
+    {
+        return SerializedQNameIterator();
+    }
+
+    return SerializedQNameIterator(BytesRange(mOutput->Buffer(), mOutput->Buffer() + mOutput->WritePos()),
+                                   mOutput->Buffer() + offset);
+}
+
+chip::Optional<uint16_t> RecordWriter::FindPreviousName(const FullQName & name) const
+{
+    for (size_t i = 0; i < kMaxCachedReferences; i++)
+    {
+        SerializedQNameIterator previous = PreviousName(i);
+
+        // Any of the sub-segments may match
+        while (previous.IsValid())
+        {
+            if (previous == name)
+            {
+                return chip::Optional<uint16_t>::Value(previous.OffsetInCurrentValidData());
+            }
+
+            if (!previous.Next())
+            {
+                break;
+            }
+        }
+    }
+
+    return chip::Optional<uint16_t>::Missing();
+}
+
+chip::Optional<uint16_t> RecordWriter::FindPreviousName(const SerializedQNameIterator & name) const
```

Comment:
It's a bit confusing why the above and this one are basically exactly the same logic. Would there be a way to template on SerializedQNameIterator/FullQName or have a base class, or something that makes the common logic factorable?

Reply:
Agree however I was unsure how to make this nicer. There  is for sure code duplication :( I will  try to see if  templates work and are readable.